### PR TITLE
fix: text with whitespace after JSX trim incorrectly

### PR DESCRIPTION
**Pull request by JounQin**
- [x] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory)
- [x] (If the change is user-facing) I’ve added my changes to the `CHANGELOG.unreleased.md` file following the template.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/master/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by JounQin** [comment]
@evilebottnawi would you like to help reviewing this PR?

**Comment by JounQin** [comment]
@ikatyang THX for reviewing, working on it.

**Comment by JounQin** [comment]
@ikatyang Fixed, please review again.

File: src/language-markdown/printer-markdown.js [hunk 1]
```diff
@@ -770,43 +770,49 @@ function isPrettierIgnore(node) {
...
 }
 
-function shouldPrePrintDoubleHardline(node, data) {
-  const isSequence = (data.prevNode && data.prevNode.type) === node.type;
-  const isSiblingNode =
-    isSequence && SIBLING_NODE_TYPES.indexOf(node.type) !== -1;
-
-  const isInTightListItem =
-    data.parentNode.type === "listItem" && !data.parentNode.loose;
+function shouldPrePrintDoubleHardline(node, { parentNode, prevNode }) {
```
Comment (ikatyang) [inline src/language-markdown/printer-markdown.js]: It'd be better to only do necessary changes or commit the refactoring/bugfix changes separately next time so that it'd be easier to review.

File: tests/mdx/__snapshots__/jsfmt.spec.js.snap [hunk 2]
```diff
@@ -247,12 +247,26 @@ printWidth: 80
...
+
+---
+
+# Heading
 <Hello>
     test   <World />   test
 </Hello>
 <Hello>
     test   <World />   test
-</Hello>123
```
Comment (ikatyang) [inline tests/mdx/__snapshots__/jsfmt.spec.js.snap]: It'd be better not to change the original test cases if it is not necessary.

File: src/language-markdown/parser-markdown.js [hunk 3]
```diff
@@ -68,13 +68,20 @@ function htmlToJsx() {
...
-        if (value) {
+        const isEl = type === "element";
+        let value = node.value.slice(
+          position.start.offset,
+          position.end.offset
+        );
+
+        if (value.trim()) {
+          const isPrevEl = newNodes[newNodes.length - 1];
+          value = isEl ? value : value.replace(/\s+/g, " "); // alike text in jsx
```
Comment (ikatyang) [inline src/language-markdown/parser-markdown.js]: Whitespaces should be processed by `splitText` in `utils.js` so you shouldn't need to process them here. And also, the newline processing should be controlled by the `--prose-wrap` option.

File: CHANGELOG.unreleased.md [hunk 4]
```diff
@@ -44,6 +44,82 @@ const link = <a href="example.com">http://example.com</a>;
...
+  test <World /> test
+</Hello>987
+
+// Output (Prettier master)
+<Hello>
+  test <World /> test
+</Hello>
+987
+
+// Input
```
[the remark this answers is not in the document]
Comment (JounQin) [reply CHANGELOG.unreleased.md]: This fix also removes redundant whitespaces, so maybe I can combine them together.

File: tests/mdx/__snapshots__/jsfmt.spec.js.snap [hunk 5]
```diff
@@ -280,6 +289,16 @@ printWidth: 80
 
 ---
 
+<Hello>
+  test <World /> test
+</Hello> 123
+
```
Comment (JounQin) [inline tests/mdx/__snapshots__/jsfmt.spec.js.snap]: Is this new line expected? @ikatyang

Comment (ikatyang) [reply tests/mdx/__snapshots__/jsfmt.spec.js.snap]: We shouldn't add a newline on L295 since it'll split the HTML (JSX) into two parts in the AST.

File: src/language-markdown/parser-markdown.js [hunk 6]
```diff
@@ -68,11 +68,12 @@ function htmlToJsx() {
...
-        const value = node.value
-          .slice(position.start.offset, position.end.offset)
-          .trim();
+        const value = node.value.slice(
+          position.start.offset,
+          position.end.offset
+        );
 
-        if (value) {
+        if (value.trim()) {
```
Comment (ikatyang) [inline src/language-markdown/parser-markdown.js]: These changes seem unnecessary.

File: tests/mdx/__snapshots__/jsfmt.spec.js.snap (hunk 5 above)
Comment (JounQin) [reply tests/mdx/__snapshots__/jsfmt.spec.js.snap]: @ikatyang Done. :)

File: src/language-markdown/parser-markdown.js [hunk 7]
```diff
@@ -68,9 +68,10 @@ function htmlToJsx() {
       }
 
       return nodes.reduce((newNodes, { sourceSpan: position, type }) => {
-        const value = node.value
-          .slice(position.start.offset, position.end.offset)
-          .trim();
+        const value = node.value.slice(
+          position.start.offset,
+          position.end.offset
+        );
```
Comment (ikatyang) [inline src/language-markdown/parser-markdown.js]: It seems you forgot to revert this change.

Comment (JounQin) [reply src/language-markdown/parser-markdown.js]: `value` should not be trimed.

**Comment by JounQin** [comment]
@lipis Can we merge this then?

**Comment by evilebottnawi** [comment]
Thanks for issue
